src/modular/boxed_monty_form.rs

```diff
@@ -82,23 +101,18 @@ impl BoxedMontyParams {
             .rem_vartime(&modulus.as_nz_ref().widen(bits_precision * 2))
             .shorten(bits_precision);
 
-        Self::new_inner(modulus, one, r2)
-    }
-
-    /// Common functionality of `new` and `new_vartime`.
-    fn new_inner(modulus: Odd<BoxedUint>, one: BoxedUint, r2: BoxedUint) -> Self {
-        debug_assert_eq!(one.bits_precision(), modulus.bits_precision());
-        debug_assert_eq!(r2.bits_precision(), modulus.bits_precision());
-
-        // If the inverse exists, it means the modulus is odd.
-        let (inv_mod_limb, modulus_is_odd) = modulus.inv_mod2k(Word::BITS);
-        debug_assert!(bool::from(modulus_is_odd));
+        // The modular inverse should always exist, because it was ensured odd above, which also ensures it's non-zero
+        let (inv_mod_limb, inv_mod_limb_exists) = modulus.inv_mod2k_full_vartime(Word::BITS);
+        debug_assert!(bool::from(inv_mod_limb_exists));
 
         let mod_neg_inv = Limb(Word::MIN.wrapping_sub(inv_mod_limb.limbs[0].0));
 
         let mod_leading_zeros = modulus.as_ref().leading_zeros().min(Word::BITS - 1);
 
-        let r3 = montgomery_reduction_boxed(&mut r2.square(), &modulus, mod_neg_inv);
+        let r3 = {
+            let mut mm = BoxedMontyMultiplier::new(&modulus, mod_neg_inv);
+            mm.square(&r2)
+        };
```

Comment:
Any reason not to just write this as...

```suggestion
        let r3 = BoxedMontyMultiplier::new(&modulus, mod_neg_inv).square(&r2);
```

Reply:
Actually, now that I think about it, we can just use normal multiplication here, since the multiplier gets discarded immediately.

Reply:
Actually number 2, we cannot use normal multiplication since the params are not created yet :) So I think the multiplier should stay, because the other option is to call `almost_montgomery_mul()` manually.